src/lib.rs

```diff
@@ -142,21 +142,67 @@ pub struct PrivateKeyRotationCheckResult {
 }
 
 impl PrivateKeyRotationCheckResult {
-    pub fn user_rotation_needed(&self) -> Option<UserId> {
+    pub fn user_rotation_needed(&self) -> Option<&UserId> {
         match &self.rotations_needed {
-            EitherOrBoth::Left(u) | EitherOrBoth::Both(u, _) => Some(u.clone()),
+            EitherOrBoth::Left(u) | EitherOrBoth::Both(u, _) => Some(u),
             _ => None,
         }
     }
 
-    pub fn group_rotation_needed(&self) -> Option<Vec1<GroupId>> {
+    pub fn group_rotation_needed(&self) -> Option<&Vec1<GroupId>> {
         match &self.rotations_needed {
-            EitherOrBoth::Right(groups) | EitherOrBoth::Both(_, groups) => Some(groups.clone()),
+            EitherOrBoth::Right(groups) | EitherOrBoth::Both(_, groups) => Some(groups),
             _ => None,
         }
     }
 
-    // pub fn rotate_all(ironoxide: &IronOxide) -> (UserId, Vec<GroupId>) //TODO consider when group private key rotation is added
+    /// Rotate the private key of the calling user and all groups they are an administrator of where needs_rotation is true.
+    /// Note that this function has the potential to take much longer than other functions, as rotation will be done
+    /// individually on each user/group. If rotation is only needed for a specific group, it is strongly recommended
+    /// to call group_rotate_private_key() instead.
+    pub fn rotate_all(
+        &self,
+        ironoxide: &IronOxide,
+        password: &str,
+    ) -> Result<(
+        Option<UserUpdatePrivateKeyResult>,
+        Option<Vec<GroupUpdatePrivateKeyResult>>,
+    )> {
+        use crate::internal::Password;
+        use futures::future::OptionFuture;
+        let valid_password: Password = password.try_into()?;
+        let user_future = self.user_rotation_needed().map(|_| {
+            crate::internal::user_api::user_rotate_private_key(
+                &ironoxide.recrypt,
+                valid_password,
+                ironoxide.device().auth(),
+            )
+        });
+        let group_futures = self.group_rotation_needed().map(|groups| {
+            let group_futures = groups
+                .into_iter()
+                .map(|group_id| {
+                    crate::internal::group_api::group_rotate_private_key(
+                        &ironoxide.recrypt,
+                        ironoxide.device().auth(),
+                        &group_id,
+                        ironoxide.device().device_private_key(),
+                    )
+                })
+                .collect::<Vec<_>>();
```

Comment:
I've noticed you having to use the turbofish quite a bit for this purpose. Note that itertools has a `collect_vec()` for this purpose. Just an option.

Reply:
I think I prefer using the turbofish since the internal type inference works well and it is consistent with the collects into non-vec types

Reply:
That's fine. Just providing another option.